File: app/src/main/java/com/drawin3d/Paper.java
package com.drawin3d;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Random;

/** Procedural sheets and surface paint. No camera, simulation loop or bitmap files. */
final class Paper {
    static final String[] KINDS={"watercolor","rough","sketch","canvas","coated"};
    static final String[] NAMES={"Watercolor · cold press","Watercolor · rough","Sketch paper","Woven canvas","Coated · water resistant"};
    static final int[] TINT={0xfff1e6d1,0xffe9dcc3,0xfff5eedf,0xffe7d8b8,0xffe4edf0};
    static final float[] SPREAD={1.35f,1.6f,1,1.08f,.92f},OPACITY={.72f,.64f,.95f,.9f,.94f},GRAIN={.55f,.85f,.22f,.6f,.08f};
    static int kind(String name){return Arrays.asList(KINDS).indexOf(name);}
    static float[][] basis(float[] normal){
        float[] n=Math3.norm(normal),u=Math3.norm(Math3.cross(Math.abs(n[1])>.95f?new float[]{0,0,-1}:new float[]{0,1,0},n));
        return new float[][]{u,Math3.cross(n,u),n};
    }
    static float[] world(SceneData.Entity e,float[] p){return Math3.add(e.position,Math3.rotateY(Math3.mul(p,e.scale),e.yaw));}
    static float[] coordinates(SceneData.Entity e,float[] world){
        float[] p=Math3.mul(Math3.rotateY(Math3.sub(world,e.position),-e.yaw),1/e.scale);float[][] axes=basis(e.normal);
        return new float[]{Math3.dot(p,axes[0]),Math3.dot(p,axes[1]),Math3.dot(p,axes[2])};
    }
    static boolean inside(SceneData.Entity e,float[] p){float[] uv=coordinates(e,p);return Math.abs(uv[0])<=e.panelWidth/2&&Math.abs(uv[1])<=e.panelWidth/e.aspect/2;}
    static int[] pixels(String kind,boolean ink){
        int k=kind(kind);int[] pixels=new int[256*256];
        for(int y=0;y<256;y++)for(int x=0;x<256;x++){
            int h=(x*374761393)^(y*668265263);h=(h^(h>>>13))*1274126177;h^=h>>>16;
            double n=(h&0xffffffffL)/4294967295.0;
            if(k==3)n=Math.min(1,n*.3+(x%8<2||y%8<2?.1:.7));
            if(k==2)n=.55+n*.35+(y%17==0?-.25:0);if(k==4)n=.88+n*.12;
            double shade=1-GRAIN[k]*(1-n)*.2;
            int r=ink?255:(int)Math.round(((TINT[k]>>16)&255)*shade),g=ink?255:(int)Math.round(((TINT[k]>>8)&255)*shade),b=ink?255:(int)Math.round((TINT[k]&255)*shade);
            int a=ink?(int)Math.round(255*(1-GRAIN[k]*(1-n)*.85)):255;pixels[y*256+x]=(a<<24)|(r<<16)|(g<<8)|b;
        }return pixels;
    }
    static Geometry sheet(SceneData.Entity e){
        Geometry m=new Geometry();float[][] axes=basis(e.normal);float w=e.panelWidth/2,h=w/e.aspect;float[] c={1,1,1,1},zero={0,0,0};
        for(int[] p:new int[][]{{-1,1},{-1,-1},{1,1},{1,1},{-1,-1},{1,-1}})m.vertex(Math3.add(Math3.mul(axes[0],p[0]*w),Math3.mul(axes[1],p[1]*h)),c,(p[0]+1)/2f,(1-p[1])/2f,zero);return m;
    }
    static Geometry ink(SceneData.Entity e,SceneData.Entity paper){return new Ink(e,paper).build();}
    private static final class Ink {
        final Geometry m=new Geometry();final SceneData.Entity e,paper;final float[][] axes;final float[] color;final int kind;final boolean wash;final Random rng;
        Ink(SceneData.Entity e,SceneData.Entity paper){this.e=e;this.paper=paper;axes=basis(paper.normal);kind=kind(paper.paperKind);wash=e.brush.equals("Water")||e.wet;rng=new Random(e.id.hashCode());color=Geometry.color(e.color,e.alpha*(wash?OPACITY[kind]:1)*(e.brush.equals("Marker")?.55f:1));}
        void vertex(float[] p,float alpha,float[] fall){
            float[] local=Math3.add(Math3.add(Math3.mul(axes[0],p[0]),Math3.mul(axes[1],p[1])),Math3.mul(axes[2],.003f/paper.scale));
            float[] ep=Math3.mul(Math3.rotateY(Math3.sub(world(paper,local),e.position),-e.yaw),1/e.scale);
            float[] fd=Math3.mul(Math3.rotateY(Math3.rotateY(Math3.add(Math3.mul(axes[0],fall[0]),Math3.mul(axes[1],fall[1])),paper.yaw),-e.yaw),paper.scale/e.scale);
            m.vertex(ep,new float[]{color[0],color[1],color[2],color[3]*alpha},.5f+(p[0]+fall[0])/paper.panelWidth,.5f-(p[1]+fall[1])/(paper.panelWidth/paper.aspect),fd);
        }
        void ribbon(float[] a,float[] b,float r,float[] fall){
            float dx=b[0]-a[0]+fall[0],dy=b[1]-a[1]+fall[1],length=(float)Math.hypot(dx,dy);if(length<1e-8f)return;
            float sx=-dy/length*r,sy=dx/length*r;float[] stops=wash?new float[]{-1,-.65f,.65f,1}:new float[]{-1,-.9f,.9f,1},opacity={0,1,1,0},zero={0,0};
            for(int i=0;i<3;i++){
                float[] a0={a[0]+sx*stops[i],a[1]+sy*stops[i]},a1={a[0]+sx*stops[i+1],a[1]+sy*stops[i+1]},b0={b[0]+sx*stops[i],b[1]+sy*stops[i]},b1={b[0]+sx*stops[i+1],b[1]+sy*stops[i+1]};
                vertex(a0,opacity[i],zero);vertex(b0,opacity[i],fall);vertex(a1,opacity[i+1],zero);vertex(a1,opacity[i+1],zero);vertex(b0,opacity[i],fall);vertex(b1,opacity[i+1],fall);
            }
        }
        void stamp(float[] p,float r){float[] zero={0,0};for(int j=0;j<12;j++){double a=j*Math.PI/6,b=(j+1)*Math.PI/6;vertex(p,1,zero);vertex(new float[]{p[0]+(float)Math.cos(a)*r,p[1]+(float)Math.sin(a)*r},0,zero);vertex(new float[]{p[0]+(float)Math.cos(b)*r,p[1]+(float)Math.sin(b)*r},0,zero);}}
        Geometry build(){
            ArrayList<float[]> all=new ArrayList<>();float[] zero={0,0};
            for(ArrayList<float[]> original:StrokePaths.of(e)){
                ArrayList<float[]> path=new ArrayList<>();for(float[] p:original){float[] uv=coordinates(paper,world(e,p));path.add(new float[]{uv[0],uv[1],p[3]});}all.addAll(path);
                for(int i=0;i<path.size();i++){float[] p=path.get(i);float r=e.width*.5f*p[2]*e.scale/paper.scale*(wash?SPREAD[kind]:1)*(e.brush.equals("Marker")?1.8f:1);
                    if(e.brush.equals("Spray")){for(int j=0;j<3;j++)stamp(new float[]{p[0]+(rng.nextFloat()-.5f)*r*5,p[1]+(rng.nextFloat()-.5f)*r*5},r*.4f);}
                    else if(i>0)ribbon(path.get(i-1),p,r,zero);else stamp(p,r);
                    if(i==path.size()-1&&path.size()>1&&!e.brush.equals("Spray"))stamp(p,r);
                }
            }
            if(e.wet&&kind==4&&!all.isEmpty()){
                float[] gravity=Math3.rotateY(new float[]{0,-1,0},-paper.yaw);float gx=Math3.dot(gravity,axes[0]),gy=Math3.dot(gravity,axes[1]);int count=Math.min(8,Math.max(1,all.size()/10));
                if(Math.hypot(gx,gy)>.05)for(int i=0;i<count;i++){float[] p=all.get(i*(all.size()-1)/Math.max(1,count-1));float length=(.08f+rng.nextFloat()*.22f)/paper.scale;ribbon(p,p,e.width*(.12f+rng.nextFloat()*.2f)*e.scale/paper.scale,new float[]{gx*length,gy*length});}
            }return m;
        }
    }
}

File: web/src/paper.js
export const PAPER_TYPES={
  watercolor:{name:'Watercolor · cold press',description:'Soft washes and fine pigment grain.',tint:0xfff1e6d1,spread:1.35,opacity:.72,grain:.55,runoff:0},
  rough:{name:'Watercolor · rough',description:'More tooth, broken colour and wider soft edges.',tint:0xffe9dcc3,spread:1.6,opacity:.64,grain:.85,runoff:0},
  sketch:{name:'Sketch paper',description:'Fine fibres and a crisp, dry mark.',tint:0xfff5eedf,spread:1,opacity:.95,grain:.22,runoff:0},
  canvas:{name:'Woven canvas',description:'A visible weave catches the paint.',tint:0xffe7d8b8,spread:1.08,opacity:.9,grain:.6,runoff:0},
  coated:{name:'Coated · water resistant',description:'Smooth surface, crisp pigment and optional running paint.',tint:0xffe4edf0,spread:.92,opacity:.94,grain:.08,runoff:1}
};
export const add=(a,b)=>a.map((v,i)=>v+b[i]),mul=(a,s)=>a.map(v=>v*s);
export const dot=(a,b)=>a[0]*b[0]+a[1]*b[1]+a[2]*b[2];
export const cross=(a,b)=>[a[1]*b[2]-a[2]*b[1],a[2]*b[0]-a[0]*b[2],a[0]*b[1]-a[1]*b[0]];
export const norm=a=>mul(a,1/Math.hypot(...a));
export function rotate(p,yaw){const a=yaw*Math.PI/180,c=Math.cos(a),s=Math.sin(a);return [c*p[0]+s*p[2],p[1],-s*p[0]+c*p[2]];}
export function basis(normal){const n=norm(normal),u=norm(cross(Math.abs(n[1])>.95?[0,0,-1]:[0,1,0],n));return {u,v:cross(n,u),n};}
export function worldPoint(e,p){return add(e.position,rotate(mul(p.slice(0,3),e.scale),e.yaw));}
// Sheet coordinates are metres on the UNROLLED sheet. Bend never rewrites ink.
export function paperAxes(paper){
  const b=basis(paper.normal),p=(paper.pitch||0)*Math.PI/180,r=(paper.roll||0)*Math.PI/180;
  const combine=q=>add(add(mul(b.u,q[0]),mul(b.v,q[1])),mul(b.n,q[2]));
  return {u:combine([Math.cos(r),Math.sin(r),0]),v:combine([-Math.sin(r)*Math.cos(p),Math.cos(r)*Math.cos(p),Math.sin(p)]),n:combine([Math.sin(r)*Math.sin(p),-Math.cos(r)*Math.sin(p),Math.cos(p)])};
}
export function bendPoint(paper,p,offset=0){const k=(paper.bend||0)*Math.PI/180/paper.panelWidth,t=p[0]*k;if(Math.abs(k)<1e-6)return [p[0],p[1],offset];return [Math.sin(t)/k-Math.sin(t)*offset,p[1],2*Math.sin(t/2)**2/k+Math.cos(t)*offset];}
export function surfaceWorld(paper,p,offset=0){const q=bendPoint(paper,p,offset),b=paperAxes(paper);return worldPoint(paper,add(add(mul(b.u,q[0]),mul(b.v,q[1])),mul(b.n,q[2])));}
export function paperLocal(paper,world){const q=mul(rotate(world.map((v,i)=>v-paper.position[i]),-paper.yaw),1/paper.scale),b=paperAxes(paper);return [dot(q,b.u),dot(q,b.v),dot(q,b.n)];}
export function paperCoordinates(paper,world){const q=paperLocal(paper,world),k=(paper.bend||0)*Math.PI/180/paper.panelWidth;return Math.abs(k)<1e-6?q:[Math.atan2(q[0]*k,1-q[2]*k)/k,q[1],0];}
export function strokeUV(e,p,paper){if(e.pointSpace!=='surface')return paperCoordinates(paper,worldPoint(e,p));const a=e.yaw*Math.PI/180,c=Math.cos(a),s=Math.sin(a);return [e.position[0]+e.scale*(p[0]*c-p[1]*s),e.position[1]+e.scale*(p[0]*s+p[1]*c),0];}
export function localizeStroke(e,paper){if(e.pointSpace==='surface')return e;return {...e,pointSpace:'surface',position:[0,0,0],normal:[0,0,1],yaw:0,width:e.width/paper.scale,points:e.points.map(p=>{const uv=strokeUV(e,p,paper);return [uv[0]/e.scale,uv[1]/e.scale,0,p[3]];})};}
export function localizePapers(entities,ids){return entities.map(e=>e.paperId&&ids.has(e.paperId)?localizeStroke(e,entities.find(p=>p.id===e.paperId)):e);}
// Analytic intersections avoid drawing on the invisible flat chord of a bent sheet.
export function hitPaper(paper,origin,direction){
  const o=paperLocal(paper,origin),end=paperLocal(paper,add(origin,direction)),d=end.map((v,i)=>v-o[i]),k=(paper.bend||0)*Math.PI/180/paper.panelWidth;
  let roots=[];
  if(Math.abs(k)<1e-6){if(Math.abs(d[2])>1e-10)roots=[-o[2]/d[2]];}
  else{const a=k*(d[0]*d[0]+d[2]*d[2]),b=2*(k*(o[0]*d[0]+o[2]*d[2])-d[2]),c=k*(o[0]*o[0]+o[2]*o[2])-2*o[2],disc=b*b-4*a*c;if(Math.abs(a)>1e-14&&disc>=0){const q=-.5*(b+Math.sign(b||1)*Math.sqrt(disc));roots=Math.abs(q)>1e-14?[q/a,c/q]:[-b/(2*a)];}}
  for(const t of roots.sort((a,b)=>a-b)){if(t<0)continue;const q=add(o,mul(d,t)),u=Math.abs(k)<1e-6?q[0]:Math.atan2(q[0]*k,1-q[2]*k)/k;if(Math.abs(u)<=paper.panelWidth/2+1e-6&&Math.abs(q[1])<=paper.panelWidth/paper.aspect/2+1e-6){const angle=u*k;if(-Math.sin(angle)*d[0]+Math.cos(angle)*d[2]>=-1e-9)return null;return {uv:[u,q[1],0],world:add(origin,mul(direction,t)),distance:t};}}
  return null;
}
// Render-only subdivision: sparse lines also follow curvature, with a fixed budget.
export function subdivideSurface(paths,paper,budget=1536){
  if(Math.abs(paper.bend||0)<1e-4)return paths;
  const total=paths.reduce((n,p)=>n+p.length,0),length=paths.reduce((n,p)=>n+p.slice(1).reduce((s,q,i)=>s+Math.abs(q[0]-p[i][0]),0),0),step=Math.max(paper.panelWidth*3/Math.max(1,Math.abs(paper.bend||0)),length/Math.max(1,budget-total));
  return paths.map(path=>path.flatMap((b,i)=>{if(!i)return [b];const a=path[i-1],n=Math.max(1,Math.ceil(Math.abs(b[0]-a[0])/step));return Array.from({length:n},(_,j)=>a.map((v,k)=>v+(b[k]-v)*(j+1)/n));}));
}
export function insidePaper(paper,world,margin=0){const [x,y]=paperCoordinates(paper,world);return Math.abs(x)<=paper.panelWidth/2-margin&&Math.abs(y)<=paper.panelWidth/paper.aspect/2-margin;}
export function attachedIds(entities,ids){const expanded=new Set(ids);for(const e of entities)if(e.paperId&&ids.has(e.paperId))expanded.add(e.id);return expanded;}
// Integer hash is identical in Java; textures are generated locally once per preset.
export function grainAt(kind,x,y){let h=Math.imul(x,374761393)^Math.imul(y,668265263);h=Math.imul(h^(h>>>13),1274126177);h^=h>>>16;const n=(h>>>0)/4294967295;
  if(kind==='canvas')return Math.min(1,n*.3+(x%8<2||y%8<2?.1:.7));
  if(kind==='sketch')return .55+n*.35+(y%17===0?-.25:0);
  if(kind==='coated')return .88+n*.12;
  return n;
}
export function paperPixels(kind,ink=false){const profile=PAPER_TYPES[kind],size=256,pixels=new Uint8ClampedArray(size*size*4);
  for(let y=0;y<size;y++)for(let x=0;x<size;x++){const i=(y*size+x)*4,n=grainAt(kind,x,y),shade=ink?1:1-profile.grain*(1-n)*.2;
    for(let k=0;k<3;k++)pixels[i+k]=Math.round((ink?255:(profile.tint>>((2-k)*8))&255)*shade);
    pixels[i+3]=ink?Math.round(255*(1-profile.grain*(1-n)*.85)):255;
  }return {pixels,size};
}
